python/src/opendp/_extrinsics/microdata/faithfulness.py

```diff
@@ -0,0 +1,127 @@
+
+from opendp.mod import Domain, Metric, Transformation
+import opendp.prelude as dp
+from opendp._lib import import_optional_dependency
+from opendp._extrinsics._utilities import to_then, with_privacy
+
+
+__all__ = ["make_faithfulness",
+           "then_faithfulness",
+           "make_private_faithfulness",
+           "then_private_faithfulness"]
+
+
+def make_faithfulness(
+        input_domain: Domain,
+        input_metric: Metric,
+        *,
+        reference_dataset,
+        similarity: Union[Callable, dict[str, float]]) -> Transformation:
+    r"""Construct a Transformation that returns the optimal (maximal) faithfulness of two datasets.
+    Faithfulness assesses whether the released dataset resemble the original dataset in a record-level granularity.
+    Namely, there is a 1-to-1 matching between the records of released dataset and the records of the original dataset: each released record is matched with an original record.
+    Two records are considered similar if their feature space distance, according to the similarity function, is less than or equal to 1.
+
+    Presented in [Israel's Birth Dataset release](https://arxiv.org/pdf/2405.00267#page=13).
+
+    :param input_domain: instance of `LazyFrameDomain`
+    :param input_metric: instance of `symmetric_distance()`
+    :param reference_dataset: public dataset to compare against, instance of `LazyFrame` or `DataFrame`
+    :param similarity: function measuing the similarity between two records
+            or a dictionary of weights for each column scaling the absolute difference
+
+    :returns a Transformation that computes a tuple of (mean, S, Vt)
+    """
+
+    np = import_optional_dependency("numpy")
+    scipy = import_optional_dependency("scipy")
+    sklearn_neighbors = import_optional_dependency("sklearn.neighbors")
+    igraph = import_optional_dependency("igraph")
+
+    dp.assert_features("contrib", "floating-point")
+
+    if "LazyFrameDomain" not in str(input_domain.type):
+        raise ValueError("input_domain must be a LazyFrame domain")
+
+    # TODO: how to enfornce public knowlege of dataset size, if at all?
+    # if (input_domain.descriptor.get("size")) is None:
+    #     raise ValueError("input_domain's size must be known")
+
+    # assert input_domain.member(reference_dataset.lazy())
+
+    if input_metric != dp.symmetric_distance():
+        raise ValueError("input metric must be symmetric distance")
+
+    if callable(similarity):
+        similarity_metric = similarity
+
+    elif isinstance(similarity, dict):
+        assert reference_dataset.columns == list(similarity.keys())
+        metric_weights = np.array(list(similarity.values()))
+
+        def similarity_metric(xi, yj):
+            return np.sum(np.abs((xi - yj)) * metric_weights)
+
+    else:
+        raise ValueError("similarity must be a callable or a dictionary")
+
+    def similarity_fn(X, Y):
+        # TODO: consider using the underline scipy implementation directly
```

Comment:
I'm really happy staying with public methods, unless there's something a lot better about the private underscore one?